proposed/2022/PrivateAssetIndepdentFromExcludeAsset.md

```diff
@@ -0,0 +1,231 @@
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+The new `PrivateAssetIndependent` property causes the `PrivateAssets` metadata to exclusively decide which assets flow to consuming parent project, but doesn't affect restore experience for current project so there would be no change in `project.assets.json` lock file, i.e it doesn't change `IncludeAssets/ExcludeAssets` calculation for current project.
+
+It'll change how `compile, runtime, contentFiles, build, buildMultitargeting, buildTransitive, analyzers, native` dependencies flow into the projects consuming it via `PackageReference` and `ProjectReference` references.
+
+For the following table assume `PrivateAssetIndependent` opt-in property is set `true` in current project, iterating possible scenarios (not full list) for consuming parent project.
+
+| Asset flowing to parent project | New feature enabled | Possible downside |
+|-----------------------|--------------|-----------------|
+| compile | Expose new code, auto complete, intellicode, intellisense | Compile error due to naming ambiguity, run time error, NU1605 error |
+| runtime | Let provide new runtime dependency | Run time error |
+| build | Enable msbuild imports | Build fails due to property/target value change |
+| analyzers | Code analyzers work | n/a |
+
+Recall our table from earlier that shows which assets flow based on listing some asset in `IncludeAssets` and/or `PrivateAssets`. Below we add a 4th column that reveals the new flow behavior when the new behavior is activated:
+
+IncludeAssets|PrivateAssets|Flows transitively (default)|Flows transitively (PrivateAssetIndependent=true)
+--|--|--|--
+☑️ | 🔲 | ✅ | ✅
+☑️ | ☑️ | ❌ | ❌
+🔲 | 🔲 | ❌ | ✅
+🔲 | ☑️ | ❌ | ❌
+
+Notice how transitive flow is controlled exclusively by the `PrivateAssets` value when `PrivateAssetIndependent` is set to `true`.
+
+#### Examples
+
+##### Case 1 for PackageReference
+
+Package reference in csproj file.
+
+```.net
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <VersionSuffix>beta</VersionSuffix>
+    <PrivateAssetIndependent>True</PrivateAssetIndependent>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" IncludeAssets="build" />
+  </ItemGroup>
+```
+
+Before change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" exclude="Runtime,Compile,Native,Analyzers,BuildTransitive" />
+      </group>
+    </dependencies>
+```
+
+After change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" include="All" />
```

Comment:
I think all the examples should use fake package IDs, like `Contoso.A`, otherwise I don't think this is a good example, because the same outcome is already achievable with `Microsoft.Windows.CsWin32` by doing `<PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" />`.

Looking at the package on nuget.org with NuGet Package Explorer, the package only contains two assets, `build` and `analyzers`. However, analyzers don't currently work via `project.assets.json`, so `IncludeAssets="build"` doesn't prevent the analyzers from being used in the current project. By removing `IncludeAssets` from the `PackageReference`, with today's NuGet it prevents all the other non-build assets being excluded in the dependency.

Reply:
Thank you for your suggestion.

> I think all the examples should use fake package IDs, like Contoso.A,

Good point. Personally, I like something tangible real examples which quickly repo and see what problem it's solving, but abstract/fake id `Contoso.A` requires previous background knowledge/experience to process what problem it's solving. What other folks are thinking about this?

> same outcome is already achievable with Microsoft.Windows.CsWin32 by doing <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" />.

Gist of this spec is both `IncludeAssets/ExcludeAssets` and `PrivateAssets` are used together scenario, not important exactly which assets are flowing transitively, in one case nothing is flowing. What other folks are thinking about this?

Reply:
I like the real package ID because of the copy-paste-testability of the spec as well.